File: bot/utils/permissions.py
"""Staff and ticket permission checks."""
import discord
from bot.database import queries as db

ADMIN_PERMISSIONS = discord.Permissions(
    manage_channels=True,
    manage_roles=True,
    moderate_members=True,
    view_channel=True,
    send_messages=True,
    manage_messages=True,
)


async def can_manage_tickets(member: discord.Member) -> bool:
    if member.guild_permissions >= ADMIN_PERMISSIONS:
        return True
    config = await db.guild_get_or_create(str(member.guild.id))
    role_ids = db.get_support_role_ids(config)
    return any(str(r.id) in role_ids for r in member.roles)


async def is_staff(member: discord.Member) -> bool:
    return await can_manage_tickets(member)


async def require_ticket_permission(member: discord.Member) -> None:
    if not await can_manage_tickets(member):
        raise PermissionError("You do not have permission to manage tickets in this server.")

File: src/utils/permissions.ts
import type { GuildMember, PermissionResolvable } from 'discord.js';
import { guildQueries } from '../database/queries.js';

const ADMIN_PERMISSIONS: PermissionResolvable = ['ManageChannels', 'ManageRoles', 'ModerateMembers', 'ViewChannel', 'SendMessages', 'ManageMessages'];

/**
 * Returns true if the member has any of the guild's configured support roles
 * or has admin-level permissions.
 */
export async function canManageTickets(member: GuildMember): Promise<boolean> {
  if (member.permissions.has(ADMIN_PERMISSIONS)) return true;
  const config = await guildQueries.getOrCreate(member.guild.id);
  const roleIds = guildQueries.getSupportRoleIds(config);
  if (roleIds.length === 0) return false;
  return roleIds.some((roleId) => member.roles.cache.has(roleId));
}

/**
 * Returns true if the member can close/reopen/escalate tickets (staff).
 */
export async function isStaff(member: GuildMember): Promise<boolean> {
  return canManageTickets(member);
}

/**
 * Assert member has ticket management permission; throws descriptive error if not.
 */
export async function requireTicketPermission(member: GuildMember): Promise<void> {
  const allowed = await canManageTickets(member);
  if (!allowed) {
    throw new Error('You do not have permission to manage tickets in this server.');
  }
}
